libraries/AP_Compass/AP_Compass_BMM150.cpp

```diff
@@ -0,0 +1,279 @@
+    _dev->write_register(OP_MODE_SELF_TEST_ODR_REG, NORMAL_MODE | ODR_30HZ);
+
+    _dev->get_semaphore()->give();
+    hal.scheduler->resume_timer_procs();
+
+    /* register the compass instance in the frontend */
+    _compass_instance = register_compass();
+    set_dev_id(_compass_instance, AP_COMPASS_TYPE_BMM150);
+
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_Compass_BMM150::_update, void));
+
+    return true;
+
+fail:
+    _dev->get_semaphore()->give();
+fail_sem:
+    hal.scheduler->resume_timer_procs();
+    return false;
+}
+
+/*
+ * Compensation algorithm got from https://github.com/BoschSensortec/BMM050_driver
+ * this is not explained in datasheet.
+ */
+int16_t AP_Compass_BMM150::_compensate_xy(int16_t xy, uint16_t rhall, bool is_x)
+{
+    int8_t txy1, txy2;
+    int32_t val;
+
+    if (!rhall) {
+        rhall = _dig_xyz1;
+    }
+
+    if (is_x) {
+        txy1 = _dig_x1;
+        txy2 = _dig_x2;
+    } else {
+        txy1 = _dig_y1;
+        txy2 = _dig_y2;
+    }
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = ((int16_t)(((uint16_t)((((int32_t)_dig_xyz1) << 14) / rhall)) -
+          ((uint16_t)0x4000)));
+    val = ((int16_t)((((int32_t)xy) * ((((((((int32_t)_dig_xy2) *
+          ((((int32_t)val) * ((int32_t)val)) >> 7)) + (((int32_t)val) *
+          ((int32_t)(((int16_t)_dig_xy1) << 7)))) >> 9) +
+          ((int32_t)0x100000)) * ((int32_t)(((int16_t)txy2) +
+          ((int16_t)0xA0)))) >> 12)) >> 13)) + (((int16_t)txy1) << 3);
+
+    return (int32_t)val;
+
+}
+
+int16_t AP_Compass_BMM150::_compensate_z(int16_t z, uint16_t rhall)
+{
+    int32_t val;
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = (((((int32_t)(z - _dig_z4)) << 15) - ((((int32_t)_dig_z3) *
+          ((int32_t)(((int16_t)rhall) - ((int16_t)_dig_xyz1)))) >> 2)) /
+          (_dig_z2 + ((int16_t)(((((int32_t)_dig_z1) *
+          ((((int16_t)rhall) << 1))) + (1 << 15)) >> 16))));
+
+    return val;
+}
+
+void AP_Compass_BMM150::_update()
+{
+    const uint32_t time_us = AP_HAL::micros();
+
+    if (time_us - _last_update_timestamp < MICRO_SECONDS_BETWEEN_MEASUREMENTS) {
+        return;
+    }
+
+    if (!_dev->get_semaphore()->take_nonblocking()) {
+        return;
+    }
+
+    uint16_t data[4];
+    _dev->read_registers(DATA_X_LSB_REG, (uint8_t *) &data, sizeof(data));
+    _dev->get_semaphore()->give();
+
+    /* Checking data ready status */
+    if (!data[3] & 0x1) {
+        return;
+    }
+
+    const uint16_t rhall = le16toh(data[3] >> 2);
+    Vector3f raw_field = Vector3f();
+
+    int16_t val = le16toh(((int16_t)data[0]) >> 3);
+    val = _compensate_xy(val, rhall, true);
+    raw_field.x = val;
+
+    val = le16toh(((int16_t)data[1]) >> 3);
+    val = _compensate_xy(val, rhall, false);
+    raw_field.y = val;
+
+    val = le16toh(((int16_t)data[2]) >> 1);
+    val = _compensate_z(val, rhall);
+    raw_field.z = val;
```

Comment:
The register 0x48 bit field is:
BIT0 = data ready status
BIT1 = fixed 0
BIT2 = RHALL LSB BIT0
BIT3 = RHALL LSB BIT1
...

Register 0x42 bit field:
BIT0 = X-Self test
BIT1 = fixed 0
BIT2 = fixed 0
BIT3 = RAW X LSB BIT0
BIT4 = RAW X LSB BIT1
...
Similar to what happen to raw Y and raw Z, that is why this shifts are done.